components/script/dom/gpubuffer.rs

```diff
@@ -0,0 +1,129 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::GPUBufferBinding::{
+    self, GPUBufferMethods, GPUBufferSize,
+};
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::reflector::DomObject;
+use crate::dom::bindings::reflector::{reflect_dom_object, Reflector};
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::window::Window;
+use dom_struct::dom_struct;
+use std::cell::Cell;
+use webgpu::{WebGPUBuffer, WebGPUDevice, WebGPURequest};
+
+#[derive(MallocSizeOf)]
+pub enum GPUBufferState {
+    Mapped,
+    Unmapped,
+    Destroyed,
+}
+
+#[dom_struct]
+pub struct GPUBuffer {
+    reflector_: Reflector,
+    label: DomRefCell<Option<DOMString>>,
+    size: Cell<GPUBufferSize>,
+    usage: Cell<u32>,
+    state: DomRefCell<GPUBufferState>,
+    buffer: WebGPUBuffer,
+    device: WebGPUDevice,
+    valid: Cell<bool>,
+}
+
+impl GPUBuffer {
+    fn new_inherited(
+        buffer: WebGPUBuffer,
+        device: WebGPUDevice,
+        state: GPUBufferState,
+        size: GPUBufferSize,
+        usage: u32,
+        valid: bool,
+    ) -> GPUBuffer {
+        Self {
+            reflector_: Reflector::new(),
+            label: DomRefCell::new(None),
+            state: DomRefCell::new(state),
+            size: Cell::new(size),
+            usage: Cell::new(usage),
+            valid: Cell::new(valid),
+            device,
+            buffer,
+        }
+    }
+
+    #[allow(unsafe_code)]
+    pub fn new(
+        global: &GlobalScope,
+        buffer: WebGPUBuffer,
+        device: WebGPUDevice,
+        state: GPUBufferState,
+        size: GPUBufferSize,
+        usage: u32,
+        valid: bool,
+    ) -> DomRoot<GPUBuffer> {
+        reflect_dom_object(
+            Box::new(GPUBuffer::new_inherited(
+                buffer, device, state, size, usage, valid,
+            )),
+            global,
+            GPUBufferBinding::Wrap,
+        )
+    }
+}
+
+impl Drop for GPUBuffer {
+    fn drop(&mut self) {
+        self.Destroy()
+    }
+}
+
+impl GPUBufferMethods for GPUBuffer {
+    /// https://gpuweb.github.io/gpuweb/#dom-gpubuffer-unmap
+    fn Unmap(&self) {
+        if let Some(window) = self.global().downcast::<Window>() {
+            match window.webgpu_channel() {
+                Some(thread) => thread
+                    .0
+                    .send(WebGPURequest::UnmapBuffer(self.buffer))
+                    .unwrap(),
+                None => unimplemented!(),
+            }
+        }
+    }
+
+    /// https://gpuweb.github.io/gpuweb/#dom-gpubuffer-destroy
+    fn Destroy(&self) {
+        match *self.state.borrow() {
+            GPUBufferState::Mapped => {
+                self.Unmap();
+            },
+            _ => {},
+        };
+        if let Some(window) = self.global().downcast::<Window>() {
```

Comment:
Since this method is called from the destructor, instead of getting the channel from the global we should hold on to the channel so we can use it directly.

Reply:
I think "holding on to the channel so we can use it directly" will be generally addressed in https://github.com/servo/servo/pull/25030, it might be a good idea to coordinate with @zakorgy and perhaps rebase this one off of those changes when they are completed(or do it in a follow-up requiring some refactoring).

Reply:
I've saved the channel in the `GPUBuffer` so we can send requests directly.